platforms/chibios/drivers/serial_protocol.c

```diff
@@ -0,0 +1,164 @@
+ */
+void soft_serial_target_init(void) {
+    driver_slave_init();
+
+    /* Start transport thread. */
+    chThdCreateStatic(waSlaveThread, sizeof(waSlaveThread), HIGHPRIO, SlaveThread, NULL);
+}
+
+/**
+ * @brief Master specific initializations.
+ */
+void soft_serial_initiator_init(void) {
+    driver_master_init();
+}
+
+/**
+ * @brief React to transactions started by the master.
+ */
+static inline bool react_to_transactions(void) {
+    /* Wait until there is a transaction for us. */
+    uint8_t sstd_index = 0;
+    receive_blocking(&sstd_index, sizeof(sstd_index));
+
+    /* Sanity check that we are actually responding to a valid transaction. */
+    if (unlikely(sstd_index >= NUM_TOTAL_TRANSACTIONS)) {
+        return false;
+    }
+
+    split_shared_memory_lock();
+    split_transaction_desc_t* trans = &split_transaction_table[sstd_index];
+
+    /* Send back the handshake which is XORed as a simple checksum,
+     to signal that the slave is ready to receive possible transaction buffers  */
+    sstd_index ^= HANDSHAKE_MAGIC;
+    if (unlikely(!send(&sstd_index, sizeof(sstd_index)))) {
+        return false;
+    }
+
+    /* Receive transaction buffer from the master. If this transaction requires it.*/
+    if (trans->initiator2target_buffer_size) {
+        if (unlikely(!receive(split_trans_initiator2target_buffer(trans), trans->initiator2target_buffer_size))) {
+            return false;
+        }
+    }
+
+    /* Allow any slave processing to occur. */
+    if (trans->slave_callback) {
+        trans->slave_callback(trans->initiator2target_buffer_size, split_trans_initiator2target_buffer(trans), trans->initiator2target_buffer_size, split_trans_target2initiator_buffer(trans));
+    }
+
+    /* Send transaction buffer to the master. If this transaction requires it. */
+    if (trans->target2initiator_buffer_size) {
+        if (unlikely(!send(split_trans_target2initiator_buffer(trans), trans->target2initiator_buffer_size))) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/**
+ * @brief Start transaction from the master half to the slave half.
+ *
+ * @param index Transaction Table index of the transaction to start.
+ * @return bool Indicates success of transaction.
+ */
+bool soft_serial_transaction(int index) {
+    split_shared_memory_lock();
+    bool result = initiate_transaction((uint8_t)index);
+    split_shared_memory_unlock();
+
+    if (unlikely(!result)) {
+        /* Clear the receive queue, to start with a clean slate.
+         * Parts of failed transactions or spurious bytes could still be in it. */
+        driver_clear();
+    }
+
+    return result;
+}
+
+/**
+ * @brief Initiate transaction to slave half.
+ */
+static inline bool initiate_transaction(uint8_t sstd_index) {
+    /* Sanity check that we are actually starting a valid transaction. */
+    if (unlikely(sstd_index >= NUM_TOTAL_TRANSACTIONS)) {
+        dprintln("USART: Illegal transaction Id.");
```

Comment:
Making the `USART`-part of the debug message configurable would  be nice, not every serial driver is a USART-driver.

Reply:
I changed it to `SPLIT` as this is generic enough in my opinion, this could be revisited in the future though when serial_protocol could become more configurable.